File: WorldFeed/Server/src/WorldFeed.Controllers/Timeline/TimelineQueryGenerator.cs
namespace WorldFeed.Controllers.Timeline
{
    using System.Text;

    using WorldFeed.Common.Extensions;
    using WorldFeed.Common.Public.Parameters.TimelineClient;
    using WorldFeed.Common.QueryGenerators;
    using WorldFeed.Common.Settings;
    using WorldFeed.Controllers.Properties;
    using WorldFeed.Controllers.Shared;

    public interface ITimelineQueryGenerator
    {
        string GetHomeTimelineQuery(IGetHomeTimelineParameters parameters, TweetMode? tweetMode);

        string GetUserTimelineQuery(IGetUserTimelineParameters parameters, TweetMode? requestTweetMode);

        // Mention Timeline
        string GetMentionsTimelineQuery(IGetMentionsTimelineParameters getMentionsTimelineParameters, TweetMode? tweetMode);

        // Retweets of Me Timeline
        string GetRetweetsOfMeTimelineQuery(IGetRetweetsOfMeTimelineParameters parameters, TweetMode? requestTweetMode);
    }

    public class TimelineQueryGenerator : ITimelineQueryGenerator
    {
        private readonly IUserQueryParameterGenerator userQueryParameterGenerator;
        private readonly IQueryParameterGenerator queryParameterGenerator;

        public TimelineQueryGenerator(IUserQueryParameterGenerator userQueryParameterGenerator, IQueryParameterGenerator queryParameterGenerator)
        {
            this.userQueryParameterGenerator = userQueryParameterGenerator;
            this.queryParameterGenerator = queryParameterGenerator;
        }

        // Home Timeline
        public string GetHomeTimelineQuery(IGetHomeTimelineParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(Resources.Timeline_GetHomeTimeline);

            this.queryParameterGenerator.AddTimelineParameters(query, parameters, requestTweetMode);

            query.AddParameterToQuery("exclude_replies", parameters.ExcludeReplies);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        // User Timeline
        public string GetUserTimelineQuery(IGetUserTimelineParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(Resources.Timeline_GetUserTimeline);

            query.AddFormattedParameterToQuery(this.userQueryParameterGenerator.GenerateIdOrScreenNameParameter(parameters.User));

            this.queryParameterGenerator.AddTimelineParameters(query, parameters, requestTweetMode);

            query.AddParameterToQuery("exclude_replies", parameters.ExcludeReplies);
            query.AddParameterToQuery("include_rts", parameters.IncludeRetweets);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        // Mentions Timeline
        public string GetMentionsTimelineQuery(IGetMentionsTimelineParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(Resources.Timeline_GetMentionsTimeline);

            this.queryParameterGenerator.AddTimelineParameters(query, parameters, requestTweetMode);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }

        // Retweets of Me Timeline
        public string GetRetweetsOfMeTimelineQuery(IGetRetweetsOfMeTimelineParameters parameters, TweetMode? requestTweetMode)
        {
            var query = new StringBuilder(Resources.Timeline_GetRetweetsOfMeTimeline);

            this.queryParameterGenerator.AddTimelineParameters(query, parameters, requestTweetMode);

            query.AddParameterToQuery("include_user_entities", parameters.IncludeUserEntities);
            query.AddFormattedParameterToQuery(parameters.FormattedCustomQueryParameters);

            return query.ToString();
        }
    }
}

File: Sharebook/Client/src/app/controllers/Timeline/TimelineQueryGenerator.ts
import {inject, Inject, Injectable, InjectionToken} from "@angular/core";

import {ComputedTweetMode} from 'src/app/core/Core/QueryGenerators/ComputedTweetMode';
import {
  IUserQueryParameterGenerator,
  IUserQueryParameterGeneratorToken
} from "../../core/Core/QueryGenerators/IUserQueryParameterGenerator";
import {Resources} from "../../properties/resources";
import {IGetHomeTimelineParameters} from "../../core/Public/Parameters/TimelineClient/GetHomeTimelineParameters";
import {IGetUserTimelineParameters} from "../../core/Public/Parameters/TimelineClient/GetUserTimelineParameters";
import {IGetMentionsTimelineParameters} from "../../core/Public/Parameters/TimelineClient/GetMentionsTimelineParameters";
import {IGetRetweetsOfMeTimelineParameters} from "../../core/Public/Parameters/TimelineClient/GetRetweetsOfMeTimelineParameters";
import {IQueryParameterGenerator, IQueryParameterGeneratorToken, QueryParameterGenerator} from "../Shared/QueryParameterGenerator";
import {UserQueryParameterGenerator} from "../User/UserQueryParameterGenerator";
import StringBuilder from "typescript-dotnet-commonjs/System/Text/StringBuilder";
import {StringBuilderExtensions} from "../../core/Core/Extensions/stringBuilder-extensions";

export interface ITimelineQueryGenerator {
  getHomeTimelineQuery(parameters: IGetHomeTimelineParameters, tweetMode: ComputedTweetMode): string;

  getUserTimelineQuery(parameters: IGetUserTimelineParameters, tweetMode: ComputedTweetMode): string;

  // Mention Timeline
  getMentionsTimelineQuery(getMentionsTimelineParameters: IGetMentionsTimelineParameters, tweetMode: ComputedTweetMode): string;

  // Retweets of Me Timeline
  getRetweetsOfMeTimelineQuery(parameters: IGetRetweetsOfMeTimelineParameters, tweetMode: ComputedTweetMode): string;
}

export const ITimelineQueryGeneratorToken = new InjectionToken<ITimelineQueryGenerator>('ITimelineQueryGenerator', {
  providedIn: 'root',
  factory: () => new TimelineQueryGenerator(inject(UserQueryParameterGenerator), inject(QueryParameterGenerator)),
});

@Injectable({
  providedIn: 'root',
})
export class TimelineQueryGenerator implements ITimelineQueryGenerator {
  private readonly _userQueryParameterGenerator: IUserQueryParameterGenerator;
  private readonly _queryParameterGenerator: IQueryParameterGenerator;

  constructor(@Inject(IUserQueryParameterGeneratorToken) userQueryParameterGenerator: IUserQueryParameterGenerator,
              @Inject(IQueryParameterGeneratorToken) queryParameterGenerator: IQueryParameterGenerator) {
    this._userQueryParameterGenerator = userQueryParameterGenerator;
    this._queryParameterGenerator = queryParameterGenerator;
  }

  // Home Timeline
  public getHomeTimelineQuery(parameters: IGetHomeTimelineParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(Resources.Timeline_GetHomeTimeline);

    this._queryParameterGenerator.addTimelineParameters(query, parameters, tweetMode);

    StringBuilderExtensions.addParameterToQuery(query, "exclude_replies", parameters.excludeReplies);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  // User Timeline
  public getUserTimelineQuery(parameters: IGetUserTimelineParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(Resources.Timeline_GetUserTimeline);

    StringBuilderExtensions.addFormattedParameterToQuery(query, this._userQueryParameterGenerator.generateIdOrScreenNameParameter(parameters.user));

    this._queryParameterGenerator.addTimelineParameters(query, parameters, tweetMode);

    StringBuilderExtensions.addParameterToQuery(query, "exclude_replies", parameters.excludeReplies);
    StringBuilderExtensions.addParameterToQuery(query, "include_rts", parameters.includeRetweets);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  // Mentions Timeline
  public getMentionsTimelineQuery(parameters: IGetMentionsTimelineParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(Resources.Timeline_GetMentionsTimeline);

    this._queryParameterGenerator.addTimelineParameters(query, parameters, tweetMode);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }

  // Retweets of Me Timeline
  public getRetweetsOfMeTimelineQuery(parameters: IGetRetweetsOfMeTimelineParameters, tweetMode: ComputedTweetMode): string {
    let query = new StringBuilder(Resources.Timeline_GetRetweetsOfMeTimeline);

    this._queryParameterGenerator.addTimelineParameters(query, parameters, tweetMode);

    StringBuilderExtensions.addParameterToQuery(query, "include_user_entities", parameters.includeUserEntities);
    StringBuilderExtensions.addFormattedParameterToQuery(query, parameters.formattedCustomQueryParameters);

    return query.toString();
  }
}
